### PR TITLE
Adding Biomass registers for ecotoplight

### PR DESCRIPTION
**Pull request by lein1013**
This should cover the changes for
https://github.com/LavermanJJ/home-assistant-solarfocus/issues/42

### CODE REVIEW

File: custom_components/solarfocus/sensor.py [hunk 1]
```diff
@@ -313,11 +314,12 @@ def native_value(self):
         key="mode",
         icon="mdi:format-list-bulleted",
         device_class=SensorDeviceClass.ENUM,
-        options=list(range(0, 3)),
+        options=list(range(0, 5)),
     ),
     SensorEntityDescription(
         key="single_charge",
         icon="mdi:pump",
+        device_class="solarfocus__bocharge"
```
Comment (LavermanJJ) [inline custom_components/solarfocus/sensor.py:322]: no required anymore as we are using `translation_key`

File: custom_components/solarfocus/sensor.py [hunk 2]
```diff
@@ -313,11 +314,12 @@ def native_value(self):
         key="mode",
         icon="mdi:format-list-bulleted",
         device_class=SensorDeviceClass.ENUM,
-        options=list(range(0, 3)),
+        options=list(range(0, 5)),
```
Comment (LavermanJJ) [inline custom_components/solarfocus/sensor.py:317]: Unrelated, but we'll keep it 👍

File: custom_components/solarfocus/sensor.py [hunk 3]
```diff
@@ -511,14 +513,16 @@ def native_value(self):
         state_class=SensorStateClass.MEASUREMENT,
     ),
     SensorEntityDescription(
-        key="status",
+        key="state",
```
Comment (LavermanJJ) [inline custom_components/solarfocus/sensor.py:516]: 🔥  breaking change ... but we'll take it for consistency

File: custom_components/solarfocus/sensor.py [hunk 4]
```diff
@@ -564,6 +568,27 @@ def native_value(self):
...
+        key="pellet_usage_total",
+        native_unit_of_measurement=UnitOfMass.KILOGRAMS,
+        icon="mdi:alpha-t-box",
+        device_class=SensorDeviceClass.WEIGHT,
+        state_class=SensorStateClass.MEASUREMENT,
+    ),
+    SensorEntityDescription(
+        key="heat_energy_total",
+        native_unit_of_measurement=UnitOfEnergy.KILO_WATT_HOUR,
+        icon="mdi:fire-circle",
```
Comment (LavermanJJ) [inline custom_components/solarfocus/sensor.py:588]:
```suggestion
        icon="mdi:meter-gas",
```
for consistency with the other thermal energy meters

File: custom_components/solarfocus/strings.json [hunk 5]
```diff
@@ -101,8 +101,73 @@
...
+          "23": "Einschubtemperatur zu hoch, Bereitschaft",
+          "24": "Lambdasonde ist defekt, Nachlauf",
+          "25": "Kesseltemperatur ist ausreichend, Bereitschaft",
+          "26": "kein Stromfluss Einschubmotor, Nachlauf",
+          "27": "Kesselfühler ist defekt, Bereitschaft",
+          "28": "Kesselfühler ist defekt, Nachlauf",
+          "29": "Lambdasonde wird beheizt",
+          "30": "Wärmetauscherreinigung wartet auf eine Freigabe, Zeit",
+          "31": "Wärmetauscherreinigung wartet auf eine Freigabe, AGT",
+          "32": " Wärmetauscherreinigung wartet auf eine Freigabe, KT",
```
Comment (LavermanJJ) [inline custom_components/solarfocus/strings.json:144]:
```suggestion
          "32": "Wärmetauscherreinigung wartet auf eine Freigabe, KT",
```
Leading space removed

File: custom_components/solarfocus/strings.json [hunk 6]
```diff
@@ -101,8 +101,73 @@
...
+          "34": "Pelletsvorratsbehälter wird befüllt",
+          "35": "Einschubfühlerkurzschluss, Bereitschaft",
+          "36": "Rückbrandschieber öffnet",
+          "37": "Kessel wird befüllt",
+          "38": "Lambdasonde wird kalibriert",
+          "39": "Kaminkehrer Messfreigabe",
+          "40": "Alarm aktiv!",
+          "41": "Kesseltür ist offen, Bereitschaft",
+          "42": "Tür war zu lange offen, Nachlauf",
+          "44": "keine Brennerzeitfreigabe, Nachlauf",
```
Comment (LavermanJJ) [inline custom_components/solarfocus/strings.json:155]:
```suggestion
          "44": "Keine Brennerzeitfreigabe, Nachlauf",
```
captialized

File: custom_components/solarfocus/strings.json [hunk 7]
```diff
@@ -101,8 +101,73 @@
...
+          "45": "Luftzahl Zündphase zu tief, Nachlauf",
+          "46": "Ascheaustragung aktiv",
+          "47": "Kesseltür/Aschebox ist offen, Bereitschaft",
+          "48": "Kesseltür/Aschebox wurde geöffnet, Nachlauf",
+          "49": "Kessel aktiv",
+          "50": "Saugaustragung wartet auf Umschalteinheit",
+          "51": "Brenner ausgeschaltet, Mindestkessellaufzeit aktiv",
+          "52": "Kaminkehrermessung beendet, Nachlauf",
+          "53": "warten bis die Raumluftklappe offen ist",
```
Comment (LavermanJJ) [inline custom_components/solarfocus/strings.json:164]:
```suggestion
          "53": "Warten bis die Raumluftklappe offen ist",
```
capitalized

File: custom_components/solarfocus/strings.json [hunk 8]
```diff
@@ -101,8 +101,73 @@
...
+          "54": "Stromausfall, Bereitschaft",
+          "55": "Stromausfall, Nachlauf",
+          "56": "WT-Spülung Brennwertmodul aktiv",
+          "57": "Differenzdruckschalter hat ausgelöst, Nachlauf",
+          "58": "warten auf Freigabe von Differenzdruckschalter",
```
Comment (LavermanJJ) [inline custom_components/solarfocus/strings.json:169]:
```suggestion
          "58": "Warten auf Freigabe von Differenzdruckschalter",
```

File: custom_components/solarfocus/strings.json [hunk 9]
```diff
@@ -200,6 +265,13 @@
           "1": "On"
         }
       },
+      "bo_charge" : {
```
Comment (LavermanJJ) [inline custom_components/solarfocus/strings.json:268]: can't find that in the spec? looks like circulation?

File: hacs.json [hunk 10]
```diff
@@ -1,9 +1,9 @@
 {
-    "name": "Solarfocus eco manager-touch",
+    "name": "Solarfocus eco manager-touch for eco top light",
```
Comment (LavermanJJ) [inline hacs.json:2]: According to how I understand Solarfocus' documentation, the eco manager-touch is included in different products [link](https://www.solarfocus.com/de/partnerportal/pdf/open/UGFydG5lcmJlcmVpY2gtREUvUmVnZWx1bmdfZWNvbWFuYWdlci10b3VjaC9BbmxlaXR1bmdlbi9lY29tYW5hZ2VyLXRvdWNoX01vZGJ1cy1UQ1AtUmVnaXN0ZXJkYXRlbl9BbmxlaXR1bmcucGRm/117920/0/Lng_YSxpM245S30zMTc4W2Y8cVRRXWlJVWRQJDsv?serialNumber=68332 )

**Review by LavermanJJ — CHANGES_REQUESTED** [review]
Thanks for providing the PR. Please see my comments.

File: custom_components/solarfocus/sensor.py (hunk 1 above)
Comment (lein1013) [reply custom_components/solarfocus/sensor.py:322]: so just remove is sufficient!?

Comment (LavermanJJ) [reply custom_components/solarfocus/sensor.py:322]: to reference in `strings.json` use `bo_single_charge`

File: custom_components/solarfocus/sensor.py (hunk 3 above)
Comment (lein1013) [reply custom_components/solarfocus/sensor.py:516]: ups, I've changed it back during testing - I'd keep status to not break it -  must be updated in parallel with pysolarfocus

File: custom_components/solarfocus/strings.json (hunk 9 above)
Comment (lein1013) [reply custom_components/solarfocus/strings.json:268]: yes, it the pump to charge the warm water boiler with the hot water from the buffer

File: hacs.json (hunk 10 above)
Comment (lein1013) [reply hacs.json:2]: Yes, the link is good - can we place it somewhere in the repository documentation?

**Comment by lein1013** [comment]
updated the PR based on the comments

File: custom_components/solarfocus/sensor.py (hunk 1 above)
Comment (lein1013) [reply custom_components/solarfocus/sensor.py]: But thats just a button -  so what then the content?
grafik

File: hacs.json (hunk 10 above)
Comment (LavermanJJ) [reply hacs.json:2]: its on the README under software versions

Comment (LavermanJJ) [reply hacs.json:2]:
```suggestion
    "name": "Solarfocus eco manager-touch",
```

File: custom_components/solarfocus/translations/en.json [hunk 11]
```diff
@@ -200,6 +265,13 @@
           "1": "On"
         }
       },
+      "bo_charge" : {
```
Comment (LavermanJJ) [inline custom_components/solarfocus/translations/en.json:268]: needs to be `bo_single_charge`

File: custom_components/solarfocus/translations/de.json [hunk 12]
```diff
@@ -199,6 +264,13 @@
           "1": "An"
         }
       },
+      "bo_charge" : {
```
Comment (LavermanJJ) [inline custom_components/solarfocus/translations/de.json:267]:
```suggestion
      "bo_single_charge" : {
```

File: custom_components/solarfocus/strings.json (hunk 9 above)
Comment (LavermanJJ) [reply custom_components/solarfocus/strings.json:268]:
```suggestion
      "bo_single_charge" : {
```

File: custom_components/solarfocus/config_flow.py [hunk 13]
```diff
@@ -45,7 +45,7 @@
 SOLARFOCUS_SYSTEMS = [
     selector.SelectOptionDict(value="Vampair", label="Heat pump vampair"),
     selector.SelectOptionDict(
-        value="Therminator", label=" Biomass boiler therminator II touch"
+        value="Therminator", label=" Biomass boiler therminator II and eco top zero light"
```
Comment (LavermanJJ) [inline custom_components/solarfocus/config_flow.py:48]:
```suggestion
        value="Therminator", label=" Biomass boiler therminator II"
```

File: custom_components/solarfocus/binary_sensor.py [hunk 14]
```diff
@@ -198,6 +198,6 @@ def is_on(self):
     SolarfocusBinarySensorEntityDescription(
         key="door_contact",
         device_class=BinarySensorDeviceClass.DOOR,
-        on_state="1",
+        on_state="0", # might be one for terminator, but is 0 for eco top light
```
Comment (LavermanJJ) [inline custom_components/solarfocus/binary_sensor.py:201]:
```suggestion
        on_state="1", 
```

File: README.md [hunk 15]
```diff
@@ -9,7 +9,8 @@ Custom component for Home-Assistant to integrate [Solarfocus](https://www.solarf
 [![hacs_badge](https://img.shields.io/badge/HACS-Default-41BDF5.svg?style=for-the-badge)](https://github.com/hacs/integration)
 
 - Install [Home Assistant Community Store (HACS)](https://hacs.xyz/)
-- Add custom repository https://github.com/lavermanjj/home-assistant-solarfocus to HACS
+- Add custom repository https://github.com/lavermanjj/home-assistant-solarfocus to HACS (original)
+- or forked version for eco top light https://github.com/lein1013/home-assistant-solarfocus
```
Comment (LavermanJJ) [inline README.md:13]:
```suggestion
```

**Review by LavermanJJ — CHANGES_REQUESTED** [review]
Please also fix the en.json and de.json translation files with the capitalizations and leading space. 
As said, I'd remove the eco-top-light support from this PR and deal with it in a separate one (for #48)

**Comment by LavermanJJ** [comment]
BTW, if you fixed review comments, feel free to also resolve the conversation.

**Comment by lein1013** [comment]
How do we deal with the hassfest check?
`for dictionary value @ data['entity']['sensor']['bo_circulation']['state']. Got {'-1': 'Locked', '0': 'Off', '1': 'On'}`

**Comment by LavermanJJ** [comment]
I guess for the moment we need to ignore them :(

File: custom_components/solarfocus/sensor.py [hunk 16]
```diff
@@ -514,11 +517,13 @@ def native_value(self):
...
         icon="mdi:fire-circle",
         device_class=SensorDeviceClass.ENUM,
-        options=list(range(200, 247)) + list(range(300, 345)),
+        options=list(range(0, 60)) + list(range(200, 247)) + list(range(300, 345)),
     ),
     SensorEntityDescription(
         key="message_number",
         icon="mdi:message-text-outline",
+        device_class=SensorDeviceClass.ENUM,
+        options=list(range(0, 2)),
```
Comment (LavermanJJ) [inline custom_components/solarfocus/sensor.py:526]: according to #59 it should be range(0,88) ?

**Review by LavermanJJ — COMMENTED** [review]
Thanks for the PR

**Pull request merged**